ingest/src/main/java/com/microsoft/azure/kusto/ingest/ResourceManager.java

```diff
@@ -346,16 +372,121 @@ protected String retrieveServiceType() {
+
+        for (int i = 0; i < retries; i++) {
+            if (i >= resources.size()) {
+                throw new IngestionClientException("Retry count exceeded");
+            }
+            TWrapper resource = resources.get(i);
+            try {
+
+                Map<String, String> attributes = new HashMap<>();
+                attributes.put("resource", resource.getEndpointWithoutSas());
+                attributes.put("account", resource.getAccountName());
+                attributes.put("type", resource.getClass().getName());
+                attributes.put("retry", String.valueOf(i));
+                return MonitoredActivity.invoke((FunctionOneException<TOut, Tracer.Span, Exception>) (Tracer.Span span) -> {
+                    try {
+                        TOut result = action.apply(resource);
+                        reportIngestionResult(resource, true);
+                        return result;
+                    } catch (Exception e) {
+                        span.addException(e);
+                        throw e;
+                    }
+                }, actionName, attributes);
+            } catch (Exception e) {
+                log.warn(String.format("Error during retry %d of %d for %s", i + 1, retries, actionName), e);
+                reportIngestionResult(resource, false);
+            }
+        }
+        throw new IngestionClientException("Retry count exceeded");
+    }
+
+    void postToQueueWithRetries(AzureStorageClient azureStorageClient, String message) throws IngestionClientException, IngestionServiceException {
+        resourceActionWithRetries(getQueues(), queue -> {
+            azureStorageClient.postMessageToQueue(queue.getQueue(), message);
+            return null;
+        }, "ResourceManager.postToQueueWithRetries");
+    }
+
+    String uploadStreamToBlobWithRetries(AzureStorageClient azureStorageClient, InputStream stream, String blobName, boolean shouldCompress)
+            throws IngestionClientException, IngestionServiceException {
+        return resourceActionWithRetries(getContainers(), container -> {
+            azureStorageClient.uploadStreamToBlob(stream, blobName, container.getContainer(), shouldCompress);
+            return (container.getContainer().getBlobContainerUrl() + "/" + blobName + container.getSas());
+        }, "ResourceManager.uploadLocalFileWithRetries");
+    }
+
+    String uploadLocalFileWithRetries(AzureStorageClient azureStorageClient, File file, String blobName, boolean shouldCompress)
+            throws IngestionClientException, IngestionServiceException {
+        return resourceActionWithRetries(getContainers(), container -> {
+            azureStorageClient.uploadLocalFileToBlob(file, blobName, container.getContainer(), shouldCompress);
+            return (container.getContainer().getBlobContainerUrl() + "/" + blobName + container.getSas());
+        }, "ResourceManager.uploadLocalFileWithRetries");
+    }
```

Comment:
This ought to be a _utils/IngestionUtils.java_ API which takes the resource manager in or a shuffled resource list and does the work.
Here it just clusters up the file.